google/cloud/storage/tests/curl_upload_request_integration_test.cc

```diff
@@ -0,0 +1,88 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/random.h"
+#include "google/cloud/log.h"
+#include "google/cloud/storage/internal/curl_request_builder.h"
+#include "google/cloud/storage/internal/nljson.h"
+#include <gmock/gmock.h>
+#include <cstdlib>
+#include <vector>
+
+using ::testing::HasSubstr;
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+
+namespace {
+std::string HttpBinEndpoint() {
+  auto env = std::getenv("HTTPBIN_ENDPOINT");
+  if (env != nullptr) {
+    return env;
+  }
+  return "https://nghttp2.org/httpbin";
+}
+
+TEST(CurlUploadRequestTest, UploadPartial) {
+  CurlRequestBuilder builder(HttpBinEndpoint() + "/post");
+  builder.AddHeader("Content-Type: application/octet-stream");
+  CurlUploadRequest upload = builder.BuildUpload();
+
+  // Create some random data to upload
+  std::string msg;
+  auto generator = google::cloud::internal::MakeDefaultPRNG();
+  std::string characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890 /\\,;-+=";
+  for (std::size_t i = 0; i != 64; ++i) {
+    msg += google::cloud::internal::Sample(generator, 127, characters);
+    msg += "\n";
+  }
+
+  std::string copy = msg;
+  auto expected_data = copy;
+
+  upload.NextBuffer(copy);
+  upload.Flush();
+  copy = msg.substr(0, 1000);
+  expected_data += copy;
+  upload.NextBuffer(copy);
+  upload.Flush();
+  copy = msg;
+  expected_data += copy;
+  upload.NextBuffer(copy);
+  copy = msg;
```

Comment:
Why no `Flush()` after the call to `NextBuffer()` (here and in the call below)?  If this is to test that `NextBuffer()` works both **with** an explicit `Flush()` call beforehand, and without one (calling `Flush()` implicitly in its method body), that's probably worth mentioning so that the next person that looks at this is aware :)